commit 0cbb58fff6fe4579279f340048cd5bf7b3033e75
Author: Todd Lipcon <todd@cloudera.com>
Date:   Fri May 29 00:29:07 2009 -0700

    HADOOP-5805. problem using top level s3 buckets as input/output directories. Contributed by Ian Nowland.
    
    Conflicts:
    
    	src/test/core/org/apache/hadoop/fs/s3native/NativeS3FileSystemContractBaseTest.java

diff --git src/core/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java src/core/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java
index 698a0c3..912ddad 100644
--- src/core/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java
+++ src/core/org/apache/hadoop/fs/s3native/NativeS3FileSystem.java
@@ -242,6 +242,11 @@ public class NativeS3FileSystem extends FileSystem {
   }
   
   private static String pathToKey(Path path) {
+    if (path.toUri().getScheme() != null && "".equals(path.toUri().getPath())) {
+      // allow uris without trailing slash after bucket to refer to root,
+      // like s3n://mybucket
+      return "";
+    }
     if (!path.isAbsolute()) {
       throw new IllegalArgumentException("Path must be absolute: " + path);
     }
diff --git src/test/org/apache/hadoop/fs/s3native/NativeS3FileSystemContractBaseTest.java src/test/org/apache/hadoop/fs/s3native/NativeS3FileSystemContractBaseTest.java
index fb867fa..a3937fd 100644
--- src/test/org/apache/hadoop/fs/s3native/NativeS3FileSystemContractBaseTest.java
+++ src/test/org/apache/hadoop/fs/s3native/NativeS3FileSystemContractBaseTest.java
@@ -61,4 +61,7 @@ public abstract class NativeS3FileSystemContractBaseTest
     assertEquals(path("/test"), paths[0].getPath());
   }
   
+  public void testNoTrailingBackslashOnBucket() throws Exception {
+    assertTrue(fs.getFileStatus(new Path(fs.getUri().toString())).isDir());
+  }
 }
